version.c

```diff
@@ -99,38 +119,45 @@ Init_version(void)
 #endif
 
 void
-Init_ruby_description(void)
+ruby_init_dynamic_description(void)
 {
-    VALUE description;
-
-    if (MJIT_OPTS_ON) {
-        description = MKSTR(description_with_mjit);
+    const size_t dynamic_description_max = strlen(ruby_description_pre) + strlen(" +MJIT +YJIT") + strlen(ruby_description_post) + 1;
+    char *new_description = ruby_xmalloc(dynamic_description_max);
+    if (strlcat(new_description, ruby_description_pre, dynamic_description_max) >= dynamic_description_max) {
+        goto description_string_overflow;
     }
-    else if (rb_yjit_enabled_p()) {
-        description = MKSTR(description_with_yjit);
+    if (MJIT_OPTS_ON) {
+        if (strlcat(new_description, " +MJIT", dynamic_description_max) >= dynamic_description_max) {
+            goto description_string_overflow;
+        }
```

Comment:
It seems that it should be possible to select between two string literals like the old code and avoid the need to use `strlcat`. Is there some particular reason you picked this approach?

Reply:
The context here is I'm trying to make it possible to add as many of these flags as we want. So three possible combinations is easy, but if I had another flag, independent of the two JIT flags, now I need six literals, etc.

The particular flag I want to add is `+MMTk` for the GC we're integrating.

I think these flags are very useful for these kind of major features, which completely change how you're debugging a crash or something, which is why we want them so prominent in the description and version string.

Reply:
Ah okay, since the badges are infix some string building seems unavoidable. I think you could make this function shorter by picking between an empty string and the `+...` string for each badge and ending with a single call to `snprintf`. You might also want to switch to a statically sized buffer instead of calling malloc. A size computed by summing `sizeof(...)` should be suitable for sizing a static character array.

Reply:
Also, since MMTK is not part of Ruby (yet), I'm tempted to keep the current simpler setup. Since MMTK will need to patch this function anyways maybe the string complexity should be in that patch instead of in this repo.